controller/src/main/java/io/pravega/controller/store/stream/records/RecordHelper.java

```diff
@@ -0,0 +1,172 @@
+
+    /**
+     * Method to compute epoch transition record. It takes segments to seal and new ranges and all the tables and
+     * computes the next epoch transition record.
+     * @param currentEpoch current epoch record
+     * @param segmentsToSeal segments to seal
+     * @param newRanges new ranges
+     * @param scaleTimestamp scale time
+     * @return new epoch transition record based on supplied input
+     */
+    public static EpochTransitionRecord computeEpochTransition(EpochRecord currentEpoch, Set<Long> segmentsToSeal,
+                                                               List<Map.Entry<Double, Double>> newRanges, long scaleTimestamp) {
+        Preconditions.checkState(segmentsToSeal.stream().allMatch(currentEpoch::containsSegment), "Invalid epoch transition request");
+
+        int newEpoch = currentEpoch.getEpoch() + 1;
+        int nextSegmentNumber = currentEpoch.getSegments().stream().mapToInt(StreamSegmentRecord::getSegmentNumber).max().getAsInt() + 1;
+        Map<Long, Map.Entry<Double, Double>> newSegments = new HashMap<>();
+        for (int i = 0; i < newRanges.size(); i++) {
+            newSegments.put(computeSegmentId(nextSegmentNumber + i, newEpoch), newRanges.get(i));
+        }
+        return new EpochTransitionRecord(currentEpoch.getEpoch(), scaleTimestamp, segmentsToSeal, newSegments);
+
+    }
+    // endregion
+
+    /**
+     * Helper method to compute list of continuous ranges. For example, two neighbouring key ranges where,
+     * range1.high == range2.low then they are considered neighbours.
+     * This method reduces input range into distinct continuous blocks.
+     * @param input list of key ranges.
+     * @return reduced list of key ranges.
+     */
+    private static List<Map.Entry<Double, Double>> reduce(List<Map.Entry<Double, Double>> input) {
+        List<Map.Entry<Double, Double>> ranges = new ArrayList<>(input);
+        ranges.sort(Comparator.comparingDouble(Map.Entry::getKey));
+        List<Map.Entry<Double, Double>> result = new ArrayList<>();
+        double low = -1.0;
+        double high = -1.0;
+        for (Map.Entry<Double, Double> range : ranges) {
+            if (high < range.getKey()) {
+                // add previous result and start a new result if prev.high is less than next.low
+                if (low != -1.0 && high != -1.0) {
+                    result.add(new AbstractMap.SimpleEntry<>(low, high));
+                }
+                low = range.getKey();
+                high = range.getValue();
+            } else if (high == range.getKey()) {
+                // if adjacent (prev.high == next.low) then update only high
+                high = range.getValue();
+            } else {
+                // if prev.high > next.low.
+                // [Note: next.low cannot be less than 0] which means prev.high > 0
+                assert low >= 0;
+                assert high > 0;
+                result.add(new AbstractMap.SimpleEntry<>(low, high));
```

Comment:
Wasn't this `<low, high>` added in the previous iteration?

Reply:
The last else is the case where there is an overlap between adjacent ranges. 
Note that low and high are included in the result set in two cases  --
1. if it is disjoint from previous range (which means they dont share a boundary) 
2. if it overlaps with previous range

`1` is handled inside if block.. 
`2` is handled inside last else block..

If they share a boundary, then the middle block `else if` handles only updating the high range. 
For example:
if we were to reduce following set:
```[0.0, 0.1) [0.1, 0.2) [0.21, 0.3) [0.25, 0.6)```
notice that it has overlaps between given ranges.
The objective is for reduction to produce following output: 
```[0.0, 0.2), [0.21, 0.3), [0.25, 0.6)```

1. iteration 1: it will enter `if` condition and set both `low` and `high` to 0.0 and 0.1 respectively but not include this range in result set yet.
so result set = {}
low = 0.0
high = 0.1
2. iteration 2: it will enter second condition (else if) which says both ranges are adjacent. 
So it will only update `high` without touching `low` and without including the value in result set. 
so result set = {}
low = 0.0
high = 0.2
3. iteration 3: it will enter first `if` block.. it will include the low to high in the result set. 
so result set = [0.0, 0.2)
low = 0.21
high = 0.3
4. iteration 4: if will enter last else block. it will include the previous range into result set and set low and high to match the new range
so result set = [0.0, 0.2), [0.21, 0.3)
low = 0.25
high = 0.6

after completing all iterations we will include the final range
so result set = [0.0, 0.2), [0.21, 0.3) [0.25, 0.6)

Reply:
I think this is a great example to add to the description on top of the method. I know this is a private method and as such it doesn't need a javadoc, but I found it illustrative.